peer_test.go

```diff
@@ -217,3 +218,120 @@ func TestPeerSelectionPreferIncoming(t *testing.T) {
 		c.Close()
 	}
 }
+
+type peerTest struct {
+	t        testing.TB
+	channels []*Channel
+}
+
+// NewService will return a new server channel and the host port.
+func (pt *peerTest) NewService(t testing.TB, svcName string) (*Channel, string) {
+	ch := testutils.NewServer(t, &testutils.ChannelOpts{ServiceName: svcName})
+	pt.channels = append(pt.channels, ch)
+	return ch, ch.PeerInfo().HostPort
+}
+
+// CleanUp will clean up all channels started as part of the peer test.
+func (pt *peerTest) CleanUp() {
+	for _, ch := range pt.channels {
+		ch.Close()
+	}
+}
+
+type peerSelectionTest struct {
+	peerTest
+
+	// numPeers is the number of peers added to the client channel.
+	numPeers int
+	// numAffinity is the number of affinity nodes.
+	numAffinity int
+	// numConcurrent is the number of concurrent goroutine to make outbound calls.
+	numConcurrent int
+	// hasInboundCall is the bool flag to tell whether to have inbound calls from affinity nodes
+	hasInboundCall bool
+
+	servers  []*Channel
+	affinity []*Channel
+	client   *Channel
+}
+
+// setupServers will create numPeer servers, and register handlers on them.
+func (pt *peerSelectionTest) setupServers(t testing.TB) {
+	pt.servers = make([]*Channel, pt.numPeers)
+
+	// Set up numPeers servers.
+	for i := 0; i < pt.numPeers; i++ {
+		pt.servers[i], _ = pt.NewService(t, "hyperbahn")
+		pt.servers[i].Register(raw.Wrap(newTestHandler(pt.t)), "echo")
+	}
+}
+
+func (pt *peerSelectionTest) setupAffinity(t testing.TB) {
+	var wg sync.WaitGroup
+	pt.affinity = make([]*Channel, pt.numAffinity)
+	for i := range pt.affinity {
+		pt.affinity[i] = pt.servers[i]
+	}
+
+	// Connect from the affinity nodes to the service.
+	for _, affinity := range pt.affinity {
+		affinity.Peers().Add(pt.client.PeerInfo().HostPort)
+		go func(affinity *Channel) {
+			pt.makeCall(affinity.GetSubChannel(pt.client.PeerInfo().ServiceName))
+			wg.Done()
+		}(affinity)
+	}
+	wg.Add(pt.numAffinity)
```

Comment:
`Add` should be called before you start any goroutines`.

What I meant by the previous comment was usually you declare the variable closest to where you use it.

You should also cache the client's host port, and avoid doing things before and in the goroutine if you can do just one of those. It should look like this:

```go

clientHP := client.PeerInfo().HostPort

var wg sync.WaitGroup
wg.Add(pt.numAffinity)
for _, affinity := range pt.affinity {
  go func(affinity *Channel) {
    defer wg.Done()

    affinity.Peers().Add(clientHP)
    // makeCall
  }(affinity)
}
```